File: Sources/Helmsman/Catalog/RandomSecret.swift
import Foundation

/// Locally-generated strong secret values (passwords, signing keys, access keys).
/// Alphanumeric only, to stay safe inside YAML scalars and shell args. Used to
/// pre-fill the install wizard's detected secret placeholders.
enum RandomSecret {
    /// Character set a generated value is drawn from. `alphanumeric` is the safe
    /// default; `hex` is needed by apps that validate a value as hex-encoded
    /// (e.g. Outline's `SECRET_KEY`, which must be `openssl rand -hex 32`).
    enum Format: String, Codable, Hashable { case alphanumeric, hex }

    private static let alphabet = Array("ABCDEFGHIJKLMNOPQRSTUVWXYZabcdefghijklmnopqrstuvwxyz0123456789")
    private static let hexAlphabet = Array("0123456789abcdef")

    static func generate(length: Int = 32, format: Format = .alphanumeric) -> String {
        let n = max(1, length)
        let chars = format == .hex ? hexAlphabet : alphabet
        var out = ""
        out.reserveCapacity(n)
        for _ in 0..<n {
            out.append(chars[Int.random(in: 0..<chars.count)])
        }
        return out
    }
}

File: packages/catalog/src/randomSecret.ts
// Locally-generated strong secret values — port of Sources/Helmsman/Catalog/RandomSecret.swift.
// Alphanumeric/hex only, to stay safe inside YAML scalars and shell args. Used
// to pre-fill the install wizard's detected secret placeholders.

import type { SecretFormat } from "./types";

const ALPHABET = "ABCDEFGHIJKLMNOPQRSTUVWXYZabcdefghijklmnopqrstuvwxyz0123456789";
const HEX_ALPHABET = "0123456789abcdef";

/** Cryptographically-random index in [0, n). Uses Web Crypto (browser + Bun). */
function randomIndex(n: number): number {
  const buf = new Uint32Array(1);
  crypto.getRandomValues(buf);
  return buf[0] % n;
}

/** Generate a random secret value. Mirrors Swift `RandomSecret.generate`. */
export function generateSecret(length = 32, format: SecretFormat = "alphanumeric"): string {
  const n = Math.max(1, length);
  const chars = format === "hex" ? HEX_ALPHABET : ALPHABET;
  let out = "";
  for (let i = 0; i < n; i++) {
    out += chars[randomIndex(chars.length)];
  }
  return out;
}
